File: src/main/java/org/tensorflow/keras/models/Sequential.java
package org.tensorflow.keras.models;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import org.tensorflow.Operand;
import org.tensorflow.Session;
import org.tensorflow.Shape;
import org.tensorflow.Tensor;
import org.tensorflow.data.GraphLoader;
import org.tensorflow.keras.callbacks.Callback;
import org.tensorflow.keras.layers.Input;
import org.tensorflow.keras.layers.Layer;
import org.tensorflow.keras.logs.BatchBeginLogs;
import org.tensorflow.keras.logs.BatchEndLogs;
import org.tensorflow.keras.logs.EpochBeginLogs;
import org.tensorflow.keras.logs.EpochEndLogs;
import org.tensorflow.keras.losses.Loss;
import org.tensorflow.keras.metrics.Metric;
import org.tensorflow.keras.optimizers.Optimizer;
import org.tensorflow.op.Ops;
import org.tensorflow.op.core.Variable;

public class Sequential<T extends Number> extends Model<T> {
    private Input<T> firstLayer;
    private Optimizer<T> optimizer;
    private List<Layer<T>> layers;

    private Loss loss;
    private List<Metric> metrics;;

    private List<Variable<T>> trainableVars;
    private List<Operand<T>> initializerOps;

    @SafeVarargs
    public Sequential(Class<T> dtype, Input<T> firstLayer, Layer<T>... layers) {
        super(dtype);
        this.firstLayer = firstLayer;
        this.layers = Arrays.asList(layers);
    }

    @SafeVarargs
    public static <T extends Number> Sequential<T> of(Class<T> dtype, Input<T> firstLayer, Layer<T>... layers) {
        return new Sequential<>(dtype, firstLayer, layers);
    }

    public Sequential addLayer(Layer<T> layer) {
        layers.add(layer);
        return this;
    }

    @Override
    public Shape computeOutputShape(Shape inputShape) {
        throw new UnsupportedOperationException("Can't call computeOutputShape on Model");
    }

    @Override
    @SafeVarargs
    public final Operand<T> call(Ops tf, Operand<T>... inputs) {
        return this.call(tf, inputs[0]);
    }

    public Operand<T> call(Ops tf, Operand<T> in) {
        Operand<T> out = in;
        for (Layer<T> layer : this.layers) {
            out = layer.apply(tf, out);
        }
        return out;
    }

    @Override
    protected void build(Ops tf, Shape inputShape) {
        throw new UnsupportedOperationException("Cannot build a sequential model");
    }

    @Override
    public void compile(Ops tf, Optimizer<T> optimizer, Loss loss, List<Metric> metrics) {
        this.loss = loss;
        this.metrics = metrics;
        this.optimizer = optimizer;

        // Targets for training loop
        this.trainableVars = new ArrayList<>();
        this.initializerOps = new ArrayList<>();

        // Build layers
        this.firstLayer.build(tf, dtype);
        Shape inputShape = firstLayer.computeOutputShape();

        for (Layer<T> layer : layers) {
            layer.build(tf, inputShape, dtype);
            this.trainableVars.addAll(layer.trainableWeights());
            this.initializerOps.addAll(layer.initializerOps());
            inputShape = layer.computeOutputShape(inputShape);
        }

        this.optimizer.build(tf, dtype);
    }

    @Override
    public void fit(Ops tf, GraphLoader<T> train, GraphLoader<T> test, int epochs, int batchSize,
            List<Callback> callbacks) {
        try (Session session = new Session(tf.scope().graph())) {
            runTrainingLoop(tf, session, train, epochs, batchSize, callbacks, true);
            runPredictionLoop(tf, session, test, batchSize, callbacks);
        }
    }

    private void runPredictionLoop(Ops tf, Session session, GraphLoader<T> data, int batchSize, List<Callback> callbacks) {
        runTrainingLoop(tf, session, data, 1, batchSize, callbacks, false);
    }

    private void runTrainingLoop(Ops tf, Session session, GraphLoader<T> data, int epochs, int batchSize,
            List<Callback> callbacks, boolean training) {
        data.batch(batchSize);
        data.build(tf);
        Operand<T>[] dataOps = data.getBatchOperands();

        Session.Runner runner;
        Operand<T> XOp = dataOps[0];
        Operand<T> yOp = dataOps[1];

        // Compute Output / Loss / Accuracy
        Operand<T> yTrue = yOp;
        Operand<T> yPred = this.apply(tf, XOp);

        Operand<T> batchLoss = loss.apply(tf, getDtype(), yTrue, yPred);
        Operand<T> batchAccuracy = this.metrics.get(0).apply(tf, getDtype(), yTrue, yPred);

        List<Operand<T>> minimize = training ? optimizer.minimize(tf, batchLoss, this.trainableVars) : null;

        for (Callback clbk : callbacks) {
            clbk.onTrainBegin();
        }

        if (training) {
            runner = session.runner();

            // Run initializer ops
            for (Operand<T> op : this.initializerOps) {
                runner.addTarget(op);
            }

            runner.run();
        }

        for (int epoch = 0; epoch < epochs; epoch++) {

            for (Callback clbk : callbacks) {
                clbk.onEpochBegin(epoch, new EpochBeginLogs());
            }

            float trainEpochAccuracy = 0;
            float trainEpochLoss = 0;

            // Load Batches
            for (int i = 0; i < data.numBatches(); i++) {

                for (Callback clbk : callbacks) {
                    BatchBeginLogs logs = new BatchBeginLogs();
                    logs.batchSize = batchSize;
                    clbk.onBatchBegin(i, logs);
                }

                runner = session.runner();
                data.feedSessionRunner(runner, i);

                if (training) {
                    for (Operand<T> op : minimize) {
                        runner.addTarget(op);
                    }
                }

                runner.fetch(batchLoss);
                runner.fetch(batchAccuracy);

                List<Tensor<?>> values = runner.run();
                try (Tensor<?> lossTensor = values.get(0); Tensor<?> accuracyTensor = values.get(1)) {
                    trainEpochAccuracy += accuracyTensor.floatValue() / data.numBatches();
                    trainEpochLoss += lossTensor.floatValue() / data.numBatches();

                    for (Callback clbk : callbacks) {
                        BatchEndLogs logs = new BatchEndLogs();
                        logs.batchAccuracy = accuracyTensor.floatValue();
                        logs.batchLoss = lossTensor.floatValue();
                        clbk.onBatchEnd(i, logs);
                    }
                }
            }

            if (training) {
                for (Callback clbk : callbacks) {
                    EpochEndLogs logs = new EpochEndLogs();
                    logs.trainAccuracy = trainEpochAccuracy;
                    logs.trainLoss = trainEpochLoss;
                    clbk.onEpochEnd(epoch, logs);
                }
            } else {
                for (Callback clbk : callbacks) {
                    EpochEndLogs logs = new EpochEndLogs();
                    logs.valAccuracy = trainEpochAccuracy;
                    logs.valLoss = trainEpochLoss;
                    clbk.onEpochEnd(epoch, logs);
                }
            }

            for (Callback clbk : callbacks) {
                clbk.onTrainEnd();
            }
        }

    }
}

File: src/main/scala/org/tensorflow/keras/models/Sequential.scala
package org.tensorflow.keras.models

import org.tensorflow.{Graph, Operand, Session, Tensor}
import org.tensorflow.data.GraphLoader
import org.tensorflow.keras.layers.{Input, Layer}
import org.tensorflow.keras.losses.Loss
import org.tensorflow.keras.metrics.Metric
import org.tensorflow.keras.optimizers.Optimizer
import org.tensorflow.ndarray.Shape
import org.tensorflow.op.Ops
import org.tensorflow.op.core.Variable
import org.tensorflow.types.TFloat32
import org.tensorflow.types.family.TNumber

import java.{util => ju}

object Sequential {
  def apply[T <: TNumber](dtype: Class[T], firstLayer: Input[T], layers: Layer[T]*): Sequential[T] =
    new Sequential[T](dtype, firstLayer, layers)
}

class Sequential[T <: TNumber](dtype: Class[T], val firstLayer: Input[T], layers0: Seq[Layer[T]])
  extends Model[T](dtype) {

  private var optimizer: Optimizer[T] = null
  final private val layers: ju.List[Layer[T]] = ju.Arrays.asList(layers0: _*)
  private var loss: Loss = null
  private var _metrics      : ju.List[Metric]       = null
  private var trainableVars : ju.List[Variable[T]]  = null
  private var initializerOpSq: ju.List[Operand [T]]  = null

  def addLayer(layer: Layer[T]): Sequential[T] = {
    layers.add(layer)
    this
  }

  override def computeOutputShape(inputShape: Shape): Shape =
    throw new UnsupportedOperationException("Can't call computeOutputShape on Model")

  final override def call(tf: Ops, inputs: Seq[Operand[T]], training: Option[Boolean]): Operand[T] =
    callOne(tf, inputs.head)

  def callOne(tf: Ops, in: Operand[T]): Operand[T] = {
    var out: Operand[T] = in

    layers.forEach { layer =>
      out = layer.apply(tf, out)
    }
    out
  }

  override protected def build(tf: Ops, inputShape: Shape): Unit =
    throw new UnsupportedOperationException("Cannot build a sequential model")

  override def compile(tf: Ops, optimizer: Optimizer[T], loss: Loss, metrics: ju.List[Metric]): Unit = {
    this.loss           = loss
    this._metrics       = metrics
    this.optimizer      = optimizer
    // Targets for training loop
    this.trainableVars  = new ju.ArrayList[Variable [T]]
    initializerOpSq = new ju.ArrayList[Operand  [T]]
    // Build layers
    this.firstLayer.build(tf, dtype)
    var inputShape: Shape = firstLayer.computeOutputShape

    layers.forEach { layer =>
      layer.build(tf, inputShape, dtype)
      this.trainableVars.addAll(layer.trainableWeights)
      initializerOpSq.addAll(layer.initializerOps)
      inputShape = layer.computeOutputShape(inputShape)
    }
    this.optimizer.build(tf, dtype)
  }

  override def fit(tf: Ops, train: GraphLoader[T], test: GraphLoader[T], epochs: Int, batchSize: Int): Unit = {
    val g: Graph = tf.scope.env.asInstanceOf[Graph] // XXX TODO ugly
    try {
      val session: Session = new Session(g)
      try {
        runTrainingLoop   (tf, session, train, epochs, batchSize, training = true)
        runPredictionLoop (tf, session, test, batchSize)
      } finally {
        session.close()
      }
    } finally {
      g.close()
    }
  }

  private def runPredictionLoop(tf: Ops, session: Session, data: GraphLoader[T], batchSize: Int): Unit =
    runTrainingLoop(tf, session, data, 1, batchSize, training = false)

  private def runTrainingLoop(tf: Ops, session: Session, data: GraphLoader[T], epochs: Int, batchSize: Int,
                              training: Boolean): Unit = {
    data.batch(batchSize)
    data.build(tf)

    val dataOps: Array[Operand[T]] = data.getBatchOperands
    var runner: Session#Runner = null
    val XOp: Operand[T] = dataOps(0)
    val yOp: Operand[T] = dataOps(1)
    // Compute Output / Loss / Accuracy
    val yTrue: Operand[T] = yOp
    val yPred: Operand[T] = this.apply(tf, XOp)

    val batchLoss     : Operand[T] = loss.apply(tf, getDtype, yTrue, yPred)
    val batchAccuracy : Operand[T] = _metrics.get(0).apply(tf, getDtype, yTrue, yPred)

    val minimize: ju.List[Operand[T]] =
      if (training) {
        optimizer.minimize(tf, batchLoss, this.trainableVars)
      } else {
        null
      }

    if (training) {
      runner = session.runner
      // Run initializer ops

      initializerOpSq.forEach { op =>
        runner.addTarget(op)
      }
      runner.run()
    }

    for (epoch <- 0 until epochs) {
      var trainEpochAccuracy = 0f
      var trainEpochLoss     = 0f
      // Load Batches
      var i = 0
      while (i < data.numBatches) {
        runner = session.runner
        data.feedSessionRunner(runner, i)
        if (training) {
          minimize.forEach { op =>
            runner.addTarget(op)
          }
        }
        runner.fetch(batchLoss)
        runner.fetch(batchAccuracy)
        val values: ju.List[Tensor] = runner.run()
        val lossTensor: TFloat32 = values.get(0).asInstanceOf[TFloat32]
        try {
          val accuracyTensor: TFloat32 = values.get(1).asInstanceOf[TFloat32]
          try {
            trainEpochAccuracy += accuracyTensor.getFloat() / data.numBatches
            trainEpochLoss     += lossTensor    .getFloat() / data.numBatches
          } finally {
            accuracyTensor.close()
          }
        } finally {
          lossTensor.close()
        }

        i += 1
      }
      if (training) {
        System.out.println("Epoch " + epoch + " train accuracy: " + trainEpochAccuracy + "  loss: " + trainEpochLoss)
      } else {
        System.out.println("Test accuracy: " + trainEpochAccuracy + " loss: " + trainEpochLoss)
      }
    }
  }
}
